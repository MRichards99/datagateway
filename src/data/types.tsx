export interface InvestigationData {
  ID: string;
  TITLE: string;
  VISIT_ID: number;
  RB_NUMBER: string;
  DOI: string;
  SIZE: number;
  INSTRUMENT: { NAME: string };
  STARTDATE: Date;
  ENDDATE: Date;
  [key: string]: string | number | Date | { NAME: string };
}

export interface DatasetData {
  ID: string;
  NAME: string;
  SIZE: number;
  CREATE_TIME: Date;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface DatafileData {
  ID: string;
  NAME: string;
  LOCATION: string;
  SIZE: number;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface FoodData {
  id: number;
  dessert: string;
  calories: number;
  fat: number;
  carbs: number;
  protein: number;
  [key: string]: string | number;
}

export type EntityType =
  | InvestigationData
  | DatasetData
  | DatafileData
  | FoodData;
