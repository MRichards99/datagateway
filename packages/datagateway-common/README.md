# datagateway-common

A library containint common code and components for the DataGateway project.
It is a dependency of other packages in the repo and other apps can import code from it
like normal.

When adding new components/functions/definitions/constants etc., you must export the item
in `index.tsx` so that it will be importable from other applications.

e.g.

In `some_file.tsx`

```
export const my_const = 1;
```

In `index.tsx`

```
export { my_const } from 'some_file.tsx'
export { my_const };
```

If you export a lot of items in `some_file.tsx` and want to export all of them to be exposed to other packages
then the following syntax can be used instead

`index.tsx`

```
export * from 'some_file.tsx'
```

## Available Scripts

In the project directory, you can run:

### `npm install`

Installs dependencies for `datagateway-common` only. Should normally use `npx lerna add` at the
top level, but can use regular `npm install` if it's only used by `datagateway-common` and not other
packages.

### `npm start`

Runs the app in development mode
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

In order to begin local development, you should uncomment the lines in `index.tsx`.
These allow the app to load like a regular React app rather than behave as a library.

### `npm run tsc`

Runs the Typescript compiler that converts our Typescript files into a Javascript library
that can be imported by our other packages. When making changes to this app, you must run
this to see the changes reflected in other apps. Think of it as a build step like `npm run build`.

### `npm test`

Runs unit tests

### `npm run test:watch`

Runs unit tests in watch mode, allowing you to filter which tests to run and automatically restart tests on code changes.

### `npm run lint:js`

Lints the `src` folder
