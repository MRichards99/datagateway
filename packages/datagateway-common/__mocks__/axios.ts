import axios from 'axios';

export default {
  get: jest.fn(() => Promise.resolve({ data: {} })),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};