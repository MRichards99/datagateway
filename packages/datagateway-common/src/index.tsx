// Uncomment to use App to test components
// import React from 'react';
// import ReactDOM from 'react-dom';
// import App from './App';

export * from './app.types';
export * from './state/actions/actions.types';

export { default as Table } from './table/table.component';
export * from './table/table.component';
export { default as DetailsPanelRow } from './table/rowRenderers/detailsPanelRow.component';
export { default as DataHeader } from './table/headerRenderers/dataHeader.component';
export { default as TextColumnFilter } from './table/columnFilters/textColumnFilter.component';
export { default as DateColumnFilter } from './table/columnFilters/dateColumnFilter.component';
export { default as ActionCellComponent } from './table/cellRenderers/actionCell.component';
export { default as DataCellComponent } from './table/cellRenderers/dataCell.component';
export { default as ExpandCellComponent } from './table/cellRenderers/expandCell.component';
export * from './table/cellRenderers/cellContentRenderers';
export * from './state/actions/datafiles';
export * from './state/actions/datasets';
export * from './state/actions/investigations';
export * from './state/actions/cart';
export * from './state/actions/instruments';
export * from './state/actions/facilityCycles';
export { default as Preloader } from './preloader/preloader.component';
export { default as dGCommonReducer } from './state/reducers/dgcommon.reducer';

import { StateType } from './state/app.types';
export type DGCommonState = StateType;

// ReactDOM.render(<App />, document.getElementById('root'));
