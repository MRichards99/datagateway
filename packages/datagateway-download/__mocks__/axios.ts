import axios from 'axios';

export default {
  get: jest.fn(path => {
    if (path.includes('/topcat/user/cart/')) {
      return Promise.resolve({ data: { cartItems: [] } });
    } else if (path === '/datagateway-download-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'LILS',
          apiUrl: 'http://scigateway-preprod.esc.rl.ac.uk:5000',
          downloadApiUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/topcat',
          idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
          accessMethods: {
            https: {
              idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
              displayName: 'HTTPS',
              description: 'Example description for HTTPS access method.',
            },
            globus: {
              idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
              displayName: 'Globus',
              description: 'Example description for Globus access method.',
            },
          },
        },
      });
    } else {
      return Promise.resolve({ data: {} });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};
