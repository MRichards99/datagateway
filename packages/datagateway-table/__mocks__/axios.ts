import axios from 'axios';

export default {
  get: jest.fn(path => {
    if (path === '/settings.json') {
      return Promise.resolve({
        data: {
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};
