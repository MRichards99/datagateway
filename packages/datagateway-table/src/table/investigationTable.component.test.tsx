import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import InvestigationTable from './investigationTable.component';
import { initialState } from '../state/reducers/dgtable.reducer';
import configureStore from 'redux-mock-store';
import { StateType } from '../state/app.types';
import {
  fetchInvestigationsRequest,
  filterTable,
  sortTable,
  addToCartRequest,
  removeFromCartRequest,
} from '../state/actions';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { TableSortLabel } from '@material-ui/core';
import { Table } from 'datagateway-common';
import { MemoryRouter } from 'react-router';
import axios from 'axios';

describe('Investigation table component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;
  (axios.get as jest.Mock).mockImplementation(() =>
    Promise.resolve({ data: [] })
  );

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'div' });
    mount = createMount();

    mockStore = configureStore([thunk]);
    state = JSON.parse(JSON.stringify({ dgtable: initialState }));
    state.dgtable.data = [
      {
        ID: 1,
        TITLE: 'Test 1',
        VISIT_ID: '1',
        RB_NUMBER: '1',
        DOI: 'doi 1',
        SIZE: 1,
        INSTRUMENT: {
          NAME: 'LARMOR',
        },
        STARTDATE: '2019-07-23',
        ENDDATE: '2019-07-24',
      },
    ];
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<InvestigationTable store={mockStore(state)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('sends fetchInvestigations action on load', () => {
    const testStore = mockStore(state);
    mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    expect(testStore.getActions()[0]).toEqual(fetchInvestigationsRequest());
  });

  it('sends filterTable action on filter', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    const filterInput = wrapper
      .find('[aria-label="Filter by Title"] input')
      .first();
    filterInput.instance().value = 'test';
    filterInput.simulate('change');

    expect(testStore.getActions()[1]).toEqual(filterTable('TITLE', 'test'));

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(testStore.getActions()[2]).toEqual(filterTable('TITLE', null));
  });

  it('sends sortTable action on sort', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find(TableSortLabel)
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(sortTable('TITLE', 'asc'));
  });

  it('sends addToCart action on unchecked checkbox click', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find('[aria-label="select row 0"]')
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(addToCartRequest());
  });

  it('sends removeFromCart action on checked checkbox click', () => {
    state.dgtable.cartItems = [
      {
        entityId: 1,
        entityType: 'investigation',
        id: 1,
        name: 'test',
        parentEntities: [],
      },
    ];

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find('[aria-label="select row 0"]')
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(removeFromCartRequest());
  });

  it('selected rows only considers relevant cart items', () => {
    state.dgtable.cartItems = [
      {
        entityId: 2,
        entityType: 'investigation',
        id: 1,
        name: 'test',
        parentEntities: [],
      },
      {
        entityId: 1,
        entityType: 'dataset',
        id: 2,
        name: 'test',
        parentEntities: [],
      },
    ];

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    const selectAllCheckbox = wrapper
      .find('[aria-label="select all rows"]')
      .first();

    expect(selectAllCheckbox.prop('checked')).toEqual(false);
    expect(selectAllCheckbox.prop('data-indeterminate')).toEqual(false);
  });

  it('renders details panel correctly', () => {
    const wrapper = shallow(
      <MemoryRouter>
        <InvestigationTable store={mockStore(state)} />
      </MemoryRouter>
    );
    const detailsPanelWrapper = shallow(
      wrapper.find(Table).prop('detailsPanel')({
        rowData: state.dgtable.data[0],
      })
    );
    expect(detailsPanelWrapper).toMatchSnapshot();
  });

  it('renders investigation title as a link', () => {
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper
        .find('[aria-colindex=3]')
        .find('p')
        .children()
    ).toMatchSnapshot();
  });

  it('renders date objects as just the date', () => {
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper
        .find('[aria-colindex=9]')
        .find('p')
        .text()
    ).toEqual('2019-07-23');

    expect(
      wrapper
        .find('[aria-colindex=10]')
        .find('p')
        .text()
    ).toEqual('2019-07-24');
  });
});
