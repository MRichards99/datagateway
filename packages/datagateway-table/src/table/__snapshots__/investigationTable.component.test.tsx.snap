// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investigation table component renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "height": 768,
      "width": "100%",
    }
  }
>
  <WithStyles(VirtualizedTable)
    columns={
      Array [
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "TITLE",
          "filterComponent": [Function],
          "label": "Title",
        },
        Object {
          "dataKey": "VISIT_ID",
          "filterComponent": [Function],
          "label": "Visit ID",
        },
        Object {
          "dataKey": "RB_NUMBER",
          "filterComponent": [Function],
          "label": "RB Number",
        },
        Object {
          "dataKey": "DOI",
          "filterComponent": [Function],
          "label": "DOI",
        },
        Object {
          "dataKey": "DATASET_COUNT",
          "label": "Dataset Count",
        },
        Object {
          "dataKey": "INSTRUMENT.NAME",
          "filterComponent": [Function],
          "label": "Instrument",
        },
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "STARTDATE",
          "label": "Start Date",
        },
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "ENDDATE",
          "label": "End Date",
        },
      ]
    }
    data={
      Array [
        Object {
          "DOI": "doi 1",
          "ENDDATE": "2019-07-24",
          "ID": 1,
          "INSTRUMENT": Object {
            "NAME": "LARMOR",
          },
          "RB_NUMBER": "1",
          "SIZE": 1,
          "STARTDATE": "2019-07-23",
          "TITLE": "Test 1",
          "VISIT_ID": "1",
        },
      ]
    }
    detailsPanel={[Function]}
    onSort={[Function]}
    sort={Object {}}
  />
</div>
`;

exports[`Investigation table component renders details panel correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Proposal:
    </b>
     
    1
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Title:
    </b>
     
    Test 1
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Start Date:
    </b>
     
    2019-07-23
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      End Date:
    </b>
     
    2019-07-24
  </WithStyles(ForwardRef(Typography))>
</div>
`;

exports[`Investigation table component renders investigation title as a link 1`] = `
<WithStyles(ForwardRef(Link))
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "Link",
      "propTypes": Object {
        "innerRef": [Function],
        "onClick": [Function],
        "replace": [Function],
        "target": [Function],
        "to": [Function],
      },
      "render": [Function],
    }
  }
  to="/browse/investigation/1/dataset"
>
  <ForwardRef(Link)
    classes={
      Object {
        "button": "MuiLink-button",
        "focusVisible": "Mui-focusVisible",
        "root": "MuiLink-root",
        "underlineAlways": "MuiLink-underlineAlways",
        "underlineHover": "MuiLink-underlineHover",
        "underlineNone": "MuiLink-underlineNone",
      }
    }
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Link",
        "propTypes": Object {
          "innerRef": [Function],
          "onClick": [Function],
          "replace": [Function],
          "target": [Function],
          "to": [Function],
        },
        "render": [Function],
      }
    }
    to="/browse/investigation/1/dataset"
  >
    <WithStyles(ForwardRef(Typography))
      className="MuiLink-root MuiLink-underlineHover"
      color="primary"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      onBlur={[Function]}
      onFocus={[Function]}
      to="/browse/investigation/1/dataset"
      variant="inherit"
    >
      <ForwardRef(Typography)
        className="MuiLink-root MuiLink-underlineHover"
        classes={
          Object {
            "alignCenter": "MuiTypography-alignCenter",
            "alignJustify": "MuiTypography-alignJustify",
            "alignLeft": "MuiTypography-alignLeft",
            "alignRight": "MuiTypography-alignRight",
            "body1": "MuiTypography-body1",
            "body2": "MuiTypography-body2",
            "button": "MuiTypography-button",
            "caption": "MuiTypography-caption",
            "colorError": "MuiTypography-colorError",
            "colorInherit": "MuiTypography-colorInherit",
            "colorPrimary": "MuiTypography-colorPrimary",
            "colorSecondary": "MuiTypography-colorSecondary",
            "colorTextPrimary": "MuiTypography-colorTextPrimary",
            "colorTextSecondary": "MuiTypography-colorTextSecondary",
            "displayBlock": "MuiTypography-displayBlock",
            "displayInline": "MuiTypography-displayInline",
            "gutterBottom": "MuiTypography-gutterBottom",
            "h1": "MuiTypography-h1",
            "h2": "MuiTypography-h2",
            "h3": "MuiTypography-h3",
            "h4": "MuiTypography-h4",
            "h5": "MuiTypography-h5",
            "h6": "MuiTypography-h6",
            "noWrap": "MuiTypography-noWrap",
            "overline": "MuiTypography-overline",
            "paragraph": "MuiTypography-paragraph",
            "root": "MuiTypography-root",
            "srOnly": "MuiTypography-srOnly",
            "subtitle1": "MuiTypography-subtitle1",
            "subtitle2": "MuiTypography-subtitle2",
          }
        }
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Link",
            "propTypes": Object {
              "innerRef": [Function],
              "onClick": [Function],
              "replace": [Function],
              "target": [Function],
              "to": [Function],
            },
            "render": [Function],
          }
        }
        onBlur={[Function]}
        onFocus={[Function]}
        to="/browse/investigation/1/dataset"
        variant="inherit"
      >
        <Link
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
          onBlur={[Function]}
          onFocus={[Function]}
          to="/browse/investigation/1/dataset"
        >
          <LinkAnchor
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="/browse/investigation/1/dataset"
            navigate={[Function]}
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="/browse/investigation/1/dataset"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Test 1
            </a>
          </LinkAnchor>
        </Link>
      </ForwardRef(Typography)>
    </WithStyles(ForwardRef(Typography))>
  </ForwardRef(Link)>
</WithStyles(ForwardRef(Link))>
`;
